pkg/upgrade/pod_restart_manager.go

```diff
@@ -0,0 +1,61 @@
+/*
+Copyright 2022 NVIDIA CORPORATION & AFFILIATES
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package upgrade
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/Mellanox/network-operator/pkg/consts"
+)
+
+// PodRestartManagerImpl implements PodRestartManager interface and can restart pods by deleting them
+type PodRestartManagerImpl struct {
+	K8sClient client.Client
+
+	Log logr.Logger
+}
+
+// PodRestartManager is and interface that allows scheduling driver pod restarts
+type PodRestartManager interface {
+	SchedulePodsRestart(context.Context, []*corev1.Pod) error
+}
+
+// SchedulePodsRestart receives a list of pods and schedules to delete them
+func (m *PodRestartManagerImpl) SchedulePodsRestart(ctx context.Context, pods []*corev1.Pod) error {
+	m.Log.V(consts.LogLevelInfo).Info("Starting Pod Restart")
+	if len(pods) == 0 {
+		m.Log.V(consts.LogLevelInfo).Info("Pod Restart Manager, no driver pods scheduled to restart")
+		return nil
+	}
+	for _, pod := range pods {
+		pod := pod
+		go func() {
```

Comment:
Maybe we can run Delete command for all PODs in the main loop? I think that with current implementation we may spawn goroutine for the same Pod more then once (new goroutine will be created by each following Reconcile cycle until Pod will be in ready state)

Reply:
Pods will only be scheduled for deletion if they are outdated and are not in "Terminating" state. The pod deletion can take quite a long time so I would prefer to still run it in a separate goroutine

Reply:
Delete call to k8s API will return immediately after DeletionTimestamp is set for resource, it will not wait for resource termination or complete removal, so it should return quite fast.

Reply:
Done